resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1

```diff
@@ -0,0 +1,297 @@
+        [string] $Version,
+        [Parameter(ValueFromPipelineByPropertyName)]
+        [bool] $PreRelease
+    )
+    
+    begin
+    {
+        function Get-DotNetToolArguments
+        {
+            param([string]$Name, [string]$Version, [bool]$PreRelease)
+
+            $string = $Name
+            
+            if (-not ([string]::IsNullOrEmpty($Version)))
+            {
+                $string += " --version $Version"
+            }
+
+            if ($PreRelease) {
+                $string += " --prerelease"
+            }
+
+            $string += " --no-cache"
+
+            return $string
+        }
+    }
+    
+    process
+    {
+        $installArgument = Get-DotNetToolArguments @PSBoundParameters
+        $arguments = "tool install $installArgument --global --ignore-failed-sources"
+        Write-Verbose -Message "Installing dotnet tool package with arguments: $arguments"
+
+        Invoke-DotNet -Command $arguments
+    }
+}
+
+function Uninstall-DotNetToolPackage
+{
+    [CmdletBinding()]
+    param (
+        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
+        [string]$Name
+    )
+
+    $arguments = "tool uninstall $Name --global" 
+    Write-Verbose -Message "Uninstalling dotnet tool package with arguments: $arguments"
+        
+    Invoke-DotNet -Command
+}
+
+function Invoke-DotNet {
+    param (
+        [Parameter(Mandatory = $true)]
+        [string] $Command
+    )
+
+    try {
+        Invoke-Expression "& `"$DotNetCliPath`" $Command"
+    }
+    catch {
+        throw "Executing dotnet.exe with {$Command} failed."
+    }
+}
+
+# Keeps the path of the code.exe CLI path.
+$DotNetCliPath = Get-DotNetPath
+
+#endregion Functions
+
+#region Classes
+[DSCResource()]
+class NETSDKToolInstaller {
+    [DscProperty(Key)]
+    [string] $PackageId
+
+    [DscProperty()]
+    [string] $Version
+
+    [DscProperty()]
+    [string[]] $Commands
+
+    [DscProperty()]
+    [bool] $PreRelease = $false
+
+    # TODO: Add support for --tool-path
+
+    [DscProperty()]
+    [bool] $Exist = $true
+
+    static [hashtable] $InstalledPackages
+
+    NETSDKToolInstaller() {
+        [NETSDKToolInstaller]::GetInstalledPackages()
+    }
+
+    NETSDKToolInstaller([string] $PackageId, [string] $Version, [string[]] $Commands, [bool] $PreRelease) {
+        $this.PackageId = $PackageId
+        $this.Version = $Version
+        $this.Commands = $Commands
+        $this.PreRelease = $PreRelease
+    }
+
+    [NETSDKToolInstaller] Get() {
+        $currentState = [NETSDKToolInstaller]::InstalledPackages[$this.PackageId]
+        if ($null -ne $currentState)
+        {
+            return $currentState
+        }
+        
+        return [NETSDKToolInstaller]@{
+            PackageId = $this.PackageId
+            Version = $this.Version
+            Commands = $this.Commands
+            Exist = $false
+        }
+    }
+
+    Set() {
+        # TODO: validate for upgrade/update scenarios
```

Comment:
what is the update model for this?

Reply:
Let's imagine the following scenario. The user has already installed PowerShell version 7.4.4. The user now wants version 7.4.5. The user leaves empty the version property. When the user runs the configuration, it needs to check if 7.4.4 is installed and if so, update it with `dotnet tool update`. @ryfu-msft Would you agree with such logic, or should we add a boolean parameter to upgrade?

Reply:
Agreed. If a package has a version that does not match the desired version, then this DSC resource should handle updating the package to match the desired state.